### PR TITLE
MINOR: random cleanup and JavaDoc improvements for clients and Streams

### PR DESCRIPTION
**Pull request by mjsax**

### CODE REVIEW

**Comment by mjsax** [comment]
Call for review @guozhangwang @dguy @bbejeck @apurvam @hachikuji

**Comment by guozhangwang** [comment]
Jenkins failure seems relevant:

```
:streams:compileTestJava/home/jenkins/jenkins-slave/workspace/kafka-pr-jdk7-scala2.11/streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.java:124: error: constructor StreamsKafkaClient in class StreamsKafkaClient cannot be applied to given types;
            super(StreamsKafkaClient.Config.fromStreamsConfig(streamsConfig), new MockClient(new MockTime()), Collections.EMPTY_LIST);
            ^
  required: Config,KafkaClient,List<MetricsReporter>,LogContext
  found: Config,MockClient,List
  reason: actual and formal argument lists differ in length
/home/jenkins/jenkins-slave/workspace/kafka-pr-jdk7-scala2.11/streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClientTest.java:216: error: constructor StreamsKafkaClient in class StreamsKafkaClient cannot be applied to given types;
        return new StreamsKafkaClient(StreamsKafkaClient.Config.fromStreamsConfig(streamsConfig),
               ^
  required: Config,KafkaClient,List<MetricsReporter>,LogContext
  found: Config,MockClient,List<MetricsReporter>
  reason: actual and formal argument lists differ in length
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
2 errors
 FAILED
```

**Comment by mjsax** [comment]
Updated.

**Comment by bbejeck** [comment]
@mjsax test failures are related the `TaskMangerTest`  checking for `restoreConsumer.assign(Collections.<TopicPartition>emptyList());` but needs to be `restoreConsumer.unsubscribe()`

**Comment by mjsax** [comment]
It is related. For `TaskMangerTest` I can simply update the test. For `StoreChangelogReaderTest` is actually exposes a buggy test that needs a proper fix.  (Btw: this issues also shows that we should not throw Java library exception in our own code base...)

**Comment by mjsax** [comment]
Updates this.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStreamThread.java [hunk 1]
```diff
@@ -247,8 +247,8 @@ void pollAndUpdate() {
 
         public void close() throws IOException {
             try {
-                consumer.close();
-            } catch (Exception e) {
+                globalConsumer.close();
+            } catch (final RuntimeException e) {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStreamThread.java:251]: Could you elaborate a bit on this change?

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 2]
```diff
@@ -711,12 +710,7 @@ private void checkBrokerVersionCompatibility() throws StreamsException {
 
         client.checkBrokerCompatibility(EXACTLY_ONCE.equals(config.getString(PROCESSING_GUARANTEE_CONFIG)));
 
-        try {
-            client.close();
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:715]: Could you elaborate a bit on this change?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClient.java [hunk 3]
```diff
@@ -280,26 +296,32 @@ private String getAnyReadyBrokerId() {
...
+    /**
+     * @return the response to the request
+     * @throws TimeoutException if there was no response within {@code request.timeout.ms}
+     * @throws StreamsException any other fatal error
+     */
+    private ClientResponse sendRequestSync(final ClientRequest clientRequest) {
         try {
             kafkaClient.send(clientRequest, Time.SYSTEM.milliseconds());
-        } catch (final Exception e) {
+        } catch (final RuntimeException e) {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClient.java:307]: Any reasons we want to change to RTE here?

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClient.java:307]: As `send()` does not throw any checked exception, it think it's cleaner code to catch `RuntimeException`.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStreamThread.java (hunk 1 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStreamThread.java:251]: As `close()` does not throw any checked exception, it think it's cleaner code to catch `RuntimeException`.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java (hunk 2 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:715]: Our internal client never throws on `close()` (cf https://github.com/apache/kafka/pull/4128/files#diff-4320e27e72244cae71428533cf3582efR173) -- no reason to catch anything.

**Comment by guozhangwang** [comment]
retest this please

**Comment by guozhangwang** [comment]
LGTM. @hachikuji do you want to make a pass over the client part?

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 4]
```diff
@@ -1385,6 +1385,10 @@ public void seekToEnd(Collection<TopicPartition> partitions) {
 
     /**
      * Get the offset of the <i>next record</i> that will be fetched (if a record with that offset exists).
+     * This method will issue a remote call to the server.
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:1388]: This method _may_ issue a remote call, but only if there is no current position for the partition.

**Review by hachikuji — COMMENTED** [review]
Just one minor comment, but looks good. I only reviewed the clients code.

**Comment by guozhangwang** [comment]
Addressed @hachikuji 's comment while merging to trunk.

**Comment by mjsax** [comment]
Thx @guozhangwang @hachikuji

**Pull request closed without merging**